quinn-proto/src/congestion/bbr.rs

```diff
@@ -0,0 +1,672 @@
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use crate::congestion::bbr_bw_estimation::BandwidthEstimation;
+use crate::congestion::PacketNumber;
+
+use super::{Controller, ControllerFactory};
+use crate::congestion::bbr_min_max::MinMax;
+use crate::connection::paths::RttEstimator;
+
+// Constants based on TCP defaults.
+// The minimum CWND to ensure delayed acks don't reduce bandwidth measurements.
+// Does not inflate the pacing rate.
+const kDefaultMinimumCongestionWindow: u64 = 4 * MAX_DATAGRAM_SIZE;
+
+// The gain used for the STARTUP, equal to 2/ln(2).
+const kDefaultHighGain: f32 = 2.885;
+// The newly derived gain for STARTUP, equal to 4 * ln(2)
+const kDerivedHighGain: f32 = 2.773;
+// The newly derived CWND gain for STARTUP, 2.
+const kDerivedHighCWNDGain: f32 = 2.0;
+// The cycle of gains used during the PROBE_BW stage.
+const kPacingGain: [f32; 8] = [1.25, 0.75, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0];
+
+const kStartupGrowthTarget: f32 = 1.25;
+const kRoundTripsWithoutGrowthBeforeExitingStartup: u8 = 3;
+
+// Do not allow initial congestion window to be greater than 200 packets.
+const kMaxInitialCongestionWindow: u64 = 200;
+
+// Do not allow initial congestion window to be smaller than 10 packets.
+const kMinInitialCongestionWindow: u64 = 10;
+
+const probe_rtt_based_on_bdp: bool = true;
+const drain_to_target: bool = true;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Mode {
+    // Startup phase of the connection.
+    STARTUP,
+    // After achieving the highest possible bandwidth during the startup, lower
+    // the pacing rate in order to drain the queue.
+    DRAIN,
+    // Cruising mode.
+    PROBE_BW,
+    // Temporarily slow down sending in order to empty the buffer and measure
+    // the real minimum RTT.
+    PROBE_RTT,
+}
+
+// Indicates how the congestion control limits the amount of bytes in flight.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum RecoveryState {
+    // Do not limit.
+    NOT_IN_RECOVERY,
+    // Allow an extra outstanding byte for each byte acknowledged.
+    CONSERVATION,
+    // Allow two extra outstanding bytes for each byte acknowledged (slow
+    // start).
+    GROWTH,
+}
+
+impl RecoveryState {
+    pub fn in_recovery(&self) -> bool {
+        !matches!(self, RecoveryState::NOT_IN_RECOVERY)
```

Comment:
Since this a "simple enum", `self != RecoveryState::NOT_IN_RECOVERY` will be fine. And clippy will probably tell you that the method should take `self` by value since its a small copyable value

Reply:
If we can avoid the need for deriving `PartialEq` (if this is the only use), I'd just as soon use `matches!()` here.